tighten_me command fails with up board image

rminnich@xcpu:~/projects/linuxboot/mainboards/aeeon/up$ utk uprom.bin tighten_me save x.bin
panic: runtime error: slice bounds out of range

goroutine 1 [running]:
github.com/linuxboot/fiano/pkg/visitors.(*TightenME).process(0xc00000c080, 0x5de8a0, 0xc00000c080)
	/home/rminnich/gopath/src/github.com/linuxboot/fiano/pkg/visitors/tightenme.go:74 +0xab4
github.com/linuxboot/fiano/pkg/visitors.(*TightenME).Run(0xc00000c080, 0x5dfde0, 0xc00008c1e0, 0x0, 0x0)
	/home/rminnich/gopath/src/github.com/linuxboot/fiano/pkg/visitors/tightenme.go:29 +0x12b
github.com/linuxboot/fiano/pkg/visitors.ExecuteCLI(0x5dfde0, 0xc00008c1e0, 0xc00000c0a0, 0x2, 0x2, 0x0, 0x0)
	/home/rminnich/gopath/src/github.com/linuxboot/fiano/pkg/visitors/cli.go:68 +0x6d
github.com/linuxboot/fiano/pkg/utk.Run(0xc000010060, 0x4, 0x4, 0x4, 0x0)
	/home/rminnich/gopath/src/github.com/linuxboot/fiano/pkg/utk/utk.go:59 +0x23f
main.main()
	/home/rminnich/gopath/src/github.com/linuxboot/fiano/cmds/utk/utk.go:75 +0x9e

Hi @JulienVdG, I did some digging and found that you can reprodruce a very similar callstack with "UP Board UEFI BIOS (UPC1DM17)" from this site https://downloads.up-community.org/download/up-board-uefi-bios-upc1dm17/
Direct link(might not be working): https://downloads.up-community.org/download/up-board-uefi-bios-upc1dm17/?wpdmdl=390

`utk UPC1DM17.bin tighten_me save x.bin
panic: runtime error: slice bounds out of range [4295045120:4190208]

goroutine 1 [running]:
github.com/linuxboot/fiano/pkg/visitors.(*TightenME).process(0xc00000c080, 0x5e9780, 0xc00000c080)
        /home/rawr/go/src/github.com/linuxboot/fiano/pkg/visitors/tightenme.go:74 +0x868
github.com/linuxboot/fiano/pkg/visitors.(*TightenME).Run(0xc00000c080, 0x5eabc0, 0xc0000941e0, 0xc000084b60, 0x0)
        /home/rawr/go/src/github.com/linuxboot/fiano/pkg/visitors/tightenme.go:29 +0xd3
github.com/linuxboot/fiano/pkg/visitors.ExecuteCLI(0x5eabc0, 0xc0000941e0, 0xc00000c0a0, 0x2, 0x2, 0x0, 0x0)
        /home/rawr/go/src/github.com/linuxboot/fiano/pkg/visitors/cli.go:68 +0x6d
github.com/linuxboot/fiano/pkg/utk.Run(0xc000010060, 0x4, 0x4, 0x4, 0x0)
        /home/rawr/go/src/github.com/linuxboot/fiano/pkg/utk/utk.go:59 +0x1db
main.main()
        /home/rawr/go/src/github.com/linuxboot/fiano/cmds/utk/utk.go:75 +0x9e`

--- a/pkg/uefi/meregion.go
+++ b/pkg/uefi/meregion.go
@@ -48,7 +48,8 @@ type MEPartitionEntry struct {
 	Owner    [4]byte
 	Offset   uint32
 	Length   uint32
-	Reserved [4]uint32
+	Reserved [3]uint32
+	Flags    uint32
 }
 
 // MEName represent 4 bytes with JSON string support
@@ -73,6 +74,22 @@ func (n *MEName) UnmarshalText(b []byte) error {
 func (n MEName) String() string {
 	b, _ := n.MarshalText()
 	return string(b)
+}
+
+// OffsetIsValid returns true if the entry has a valid offset
+func (e MEPartitionEntry) OffsetIsValid() bool {
+	return e.Offset != 0 && e.Offset != 0xffffffff
+}
+
+var mePartitionEntryTypeNames = map[byte]string{0: "Code", 1: "Data", 2: "NVRAM", 3: "Generic", 4: "EFFS", 5: "ROM"}
+
+// Type returns the type of the entry
+func (e MEPartitionEntry) Type() string {
+	t := byte(e.Flags & 0x7f)
+	if s, ok := mePartitionEntryTypeNames[t]; ok {
+		return s
+	}
+	return fmt.Sprintf("Unknown (%d)", t)
 }
 
 // FindMEDescriptor searches for an Intel ME FTP signature
@@ -182,9 +199,11 @@ func NewMERegion(buf []byte, r *FlashRegion, rt FlashRegionType) (Region, error)
 	rr.FPT = fp
 	// Compute FreeSpaceOffset
 	for _, p := range fp.Entries {
-		endOffset := uint64(p.Offset) + uint64(p.Length)
-		if endOffset > rr.FreeSpaceOffset {
-			rr.FreeSpaceOffset = endOffset
+		if p.OffsetIsValid() {
+			endOffset := uint64(p.Offset) + uint64(p.Length)
+			if endOffset > rr.FreeSpaceOffset {
+				rr.FreeSpaceOffset = endOffset
+			}
 		}
 	}
 

--- a/pkg/visitors/table.go
+++ b/pkg/visitors/table.go
@@ -86,7 +86,7 @@ func (v *Table) printFirmware(f uefi.Firmware, node, name, typez interface{}, of
 		v.W = tabwriter.NewWriter(os.Stdout, 0, 0, 2, ' ', 0)
 		defer func() { v.W.Flush() }()
 		if v.Layout {
-			fmt.Fprintf(v.W, "%sNode\tGUID/Name\tOffset\tSize\n", indent(v.indent))
+			fmt.Fprintf(v.W, "%sNode\tGUID/Name/Type\tOffset\tSize\n", indent(v.indent))
 			v.printRow = printRowLayout
 		} else {
 			fmt.Fprintf(v.W, "%sNode\tGUID/Name\tType\tSize\n", indent(v.indent))
@@ -129,7 +129,11 @@ func (v *Table) printFirmware(f uefi.Firmware, node, name, typez interface{}, of
 	case *uefi.MEFPT:
 		// MERegion is not entered, simply print the $FPT content here
 		for _, p := range f.Entries {
-			v2.printRow(&v2, p.Name, "", "", offset+uint64(p.Offset), uint64(p.Length))
+			var po uint64
+			if p.OffsetIsValid() {
+				po = offset + uint64(p.Offset)
+			}
+			v2.printRow(&v2, p.Name, "", p.Type(), po, uint64(p.Length))
 		}
 	case *uefi.File:
 		// Align

